### PR TITLE
docs: mention equivalent config syntaxes

### PR DESCRIPTION
**Pull request by thadguidry**
## Pre-flight checklist

- [X] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [ ] **If this is a code change**: I have written unit tests and/or added dogfooding pages to fully verify the new behavior.
- [ ] **If this is a new API or substantial change**: the PR has an accompanying issue (closes #0000) and the maintainers have approved on my working plan.

## Motivation

Fixes part of #8925

### Test links

https://deploy-preview-8951--docusaurus-2.netlify.app/docs/configuration/#syntax-to-declare-docusaurus-config

https://deploy-preview-8951--docusaurus-2.netlify.app/docs/api/docusaurus-config/

## Related issues/PRs

#8925

### CODE REVIEW

File: website/docs/api/docusaurus.config.js.mdx [hunk 1]
```diff
@@ -6,6 +6,12 @@ slug: /api/docusaurus-config
 
 # `docusaurus.config.js`
 
+:::info
+
+Refer to the Getting Started [**Configuration**](docs/configuration.mdx) for examples.
+
+:::
+
```
Comment (slorber) [inline website/docs/api/docusaurus.config.js.mdx:14]: all these things were already in the other PR so shouldn't be there

File: website/docs/configuration.mdx [hunk 2]
```diff
@@ -49,6 +54,14 @@ module.exports = {
...
 ```
 
+:::info[About the default `docusaurus.config.js` template]
+
+The init template `docusaurus.config.js` uses the `const config = { ... }` and it is equivalent to `module.exports = { ... }` shown throughout the examples.
+There are various equivalent ways to declare your Docusaurus config, such as using creator functions and async code.
+See: [**`docusaurus.config.js` API reference**](api/docusaurus.config.js.mdx)
+
+:::
+
```
Comment (slorber) [inline website/docs/configuration.mdx:64]: This is in the `### Theme, plugin, and preset configurations` section, and does not look like an appropriate place to me

I'd suggest to create a paragraph `## Syntax to declare docusaurus.config.js`, and add more content here, including multiple non-inline code examples, including async examples currently found in the API ref, link to nodejs doc about import/require etc

**Review by slorber — CHANGES_REQUESTED** [review]

File: website/docs/configuration.mdx (hunk 2 above)
Comment (thadguidry) [reply website/docs/configuration.mdx:64]: Which doc link is more important for nodejs import/require?  I ask because is `docusaurus.config.js` expecting only CommonJS or utilizing ECMAScript module loading and has interoperability with CommonJS?
https://nodejs.org/api/esm.html#modules-ecmascript-modules
and
https://nodejs.org/api/esm.html#interoperability-with-commonjs

I see the compiler target itself is ES2020?
```json
{
  "compilerOptions": {
    /* Emit */
    "target": "ES2020",
```

File: website/docs/api/docusaurus.config.js.mdx (hunk 1 above)
Comment (thadguidry) [reply website/docs/api/docusaurus.config.js.mdx:14]: So just this refer sentence and link is good enough here as is now, correct?  You did say you wanted to link back from the API ref to the docs/guide for configuration.  So that is what this info admonition is.

File: website/docs/configuration.mdx [hunk 3]
```diff
@@ -20,7 +97,6 @@ The high-level overview of Docusaurus configuration can be categorized into:
 
 <TOCInline toc={toc} minHeadingLevel={3} maxHeadingLevel={3} />
 
-For exact reference to each of the configurable fields, you may refer to [**`docusaurus.config.js` API reference**](api/docusaurus.config.js.mdx).
```
Comment (slorber) [inline website/docs/configuration.mdx:23]: not needed, lead to git conflicts

File: website/docs/configuration.mdx [hunk 4]
```diff
@@ -49,6 +125,8 @@ module.exports = {
 };
 ```
 
+
+
```
Comment (slorber) [inline website/docs/configuration.mdx:129]: ? linter won't like it

File: website/docs/configuration.mdx [hunk 5]
```diff
@@ -6,10 +6,87 @@ description: Configuring your site's behavior through docusaurus.config.js and m
...
+Check the [**`docusaurus.config.js` API reference**](api/docusaurus.config.js.mdx) for an exhaustive list of options.
+
+:::
+
 Docusaurus has a unique take on configurations. We encourage you to congregate information about your site into one place. We guard the fields of this file and facilitate making this data object accessible across your site.
 
 Keeping a well-maintained `docusaurus.config.js` helps you, your collaborators, and your open source contributors to be able to focus on documentation while still being able to customize the site.
 
+## Syntax to declare docusaurus.config.js
```
Comment (slorber) [inline website/docs/configuration.mdx:19]:
```suggestion
## Syntax to declare `docusaurus.config.js` {#syntax-to-declare-docusaurus-config}
```

File: website/docs/configuration.mdx [hunk 6]
```diff
@@ -6,10 +6,87 @@ description: Configuring your site's behavior through docusaurus.config.js and m
...
+```js title="docusaurus.config.js"
+module.exports = Promise.resolve({
+  // site config...
+});
+
+```
+
+Since a config file is Javascript then import/require can be utilized to pull in libraries to help.  (For advanced users, just know that we )
+
+// TODO: link for nodejs import/require.  And answer question about interoperability for CommonJS and ECMAScript modules.
```
Comment (slorber) [inline website/docs/configuration.mdx:88]: That's too verbose

What I want is something concice with many small examples like this

CleanShot 2023-05-04 at 11 21 55@2x

**Review by slorber — CHANGES_REQUESTED** [review]

File: website/docs/configuration.mdx [hunk 7]
```diff
@@ -6,10 +6,87 @@ description: Configuring your site's behavior through docusaurus.config.js and m
 
 import TOCInline from '@theme/TOCInline';
 
+:::info
+
+Check the [**`docusaurus.config.js` API reference**](api/docusaurus.config.js.mdx) for an exhaustive list of options.
+
+:::
+
```
Comment (slorber) [inline website/docs/configuration.mdx:14]: we already merged that in the other PR so it shouldn't be as well in this PR

File: website/docs/configuration.mdx (hunk 2 above)
Comment (slorber) [reply website/docs/configuration.mdx]: We don't support ES modules, so my mistake we shouldn't link to any doc mentioning imports or this will be confusing. (including the Node.js docs unfortunately, hard to follow, not very well presented and also mentioning ES modules)

It is surprisingly hard to find easy-to-understand explanations of CommonJS modules. 
This one is pretty simple: https://flaviocopes.com/commonjs/

You can say something like:

> The Docusaurus config file only supports the [CommonJS](https://flaviocopes.com/commonjs/) module system (`require("lib")` / `module.exports = /* your config*/`).

**Comment by thadguidry** [comment]
To me, this now looks and feels really good.  The missing info and examples that would have simply explained things for me at the very beginning of my confusion and opening issue #8925.  I hope it now matches all our expectations for more clear docs for novice users trying out Docusaurus for the first time.

@slorber so, what do you think now?

**Comment by slorber** [comment]
Thanks, will do some minor edits and merge
